exercises/tree_iterators/descr.md

```diff
@@ -0,0 +1,180 @@
+
+A type `'a tree` of binary trees is given, where every binary node is
+decorated with an *element*, that is, a piece of data of type `'a`.
+This type is defined as follows:
+
+```
+type 'a tree =
+  | Leaf
+  | Node of 'a tree * 'a * 'a tree
+```
+
+One could impose invariants on this data structure, including an ordering
+invariant and a balancing invariant, to obtain *balanced binary search trees*,
+a simple, widely-known implementation of dictionaries. In this exercise,
+however, these invariants play no role.
+
+The **fringe** of a tree is the list of its elements, in the order in which
+they appear when the tree is displayed as an OCaml expression. (This is
+sometimes known as an infix enumeration.) For instance, the fringe of the tree
+`Node (Node (Leaf, 0, Leaf), 1, Node (Leaf, 2, Leaf))` is the list
+`[0; 1; 2]`.
+
+## Building the fringe of a tree (eagerly)
+
+**Question 1.**
+Define a function `slow_elements`
+of type `'a tree -> 'a list`
+such that `slow_elements t` is the fringe of the tree `t`.
+Give a direct recursive definition of this function,
+aiming for simplicity, not for efficiency.
+
+When Question 1 above is answered in the most straightforward way,
+the solution involves the list concatenation function `(@)`, whose
+time complexity is linear in the length of its first argument.
+As a result, the complexity of the function `slow_elements` is bad:
+it is `O(n^2)` on arbitrary trees,
+and `O(n log n)` on weight-balanced trees.
+Yet, ideally, traversing a tree should take only linear time.
+
+A classic solution to this problem is to
+write a function `elements_with`
+that takes a **continuation list** `ys` as
+a second argument. Each tree element is
+consed (that is, prepended) to `ys`,
+thereby eliminating the need for list concatenation.
+
+**Question 2.**
+Define a function `elements_with`
+of type `'a tree -> 'a list -> 'a list`
+such that the list `elements_with t ys` is
+equal to the list `slow_elements t @ ys`.
+
+**Question 3.**
+In one line,
+define a function `elements`
+of type `'a tree -> 'a list`
+such that `elements t` is the fringe of the tree `t`.
+(Thus, the lists `elements t` and `slow_elements t` are equal.)
+
+The function `elements` obtained in Question 3 builds the fringe of a tree in
+linear time, which is good. A problem remains, however: suppose that someone
+needs access to just the first `k` elements of the fringe, but the value of
+`k` is not known ahead of time. In such a scenario, building the entire fringe
+in memory is wasteful. We would like to be able to produce elements **on
+demand**, with `O(1)` work per element, so that the first `k` elements are
+produced in time `O(k)`.
+
+Fortunately, there is a very simple way of doing so. The trick is to use a
+slightly more elaborate, yet closely related data structure: we replace
+*lists* with *on-demand sequences*.
+
+## On-demand sequences
+
+The type of **on-demand sequences** is defined in a module named `Seq`.
+Beginning with version 4.07,
```

Comment:
So I understand why `ocaml 4.05` is not enough for your needs. I think you can assume that learn-ocaml will be synchronized with the latest version of OCaml in the future.

Reply:
OK. That said, the `Seq` module is OCaml 4.07 is very basic; it does not even have `Seq.nil` and `Seq.cons`, which are useful here, so I may end up keeping my little `Seq` module in `prelude.ml` the way it is now.